File: app/src/main/java/com/example/projectquestonjava/core/managers/DataStoreManager.java
package com.example.projectquestonjava.core.managers;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.datastore.preferences.core.MutablePreferences;
import androidx.datastore.preferences.core.Preferences;
import androidx.datastore.rxjava3.RxDataStore;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.LiveDataReactiveStreams;
import com.example.projectquestonjava.core.di.IODispatcher;
import com.example.projectquestonjava.core.utils.Logger;
import com.google.common.util.concurrent.ListenableFuture;
import com.google.common.util.concurrent.SettableFuture;
import java.io.IOException;
import java.util.NoSuchElementException;
import java.util.Objects;
import java.util.concurrent.Executor;
import javax.inject.Inject;
import javax.inject.Singleton;
import io.reactivex.rxjava3.core.Flowable;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.disposables.Disposable;
import io.reactivex.rxjava3.schedulers.Schedulers;

@Singleton
public class DataStoreManager {

    private static final String TAG = "DataStoreManager";
    private final RxDataStore<Preferences> dataStore;
    private final Logger logger;
    private final Executor ioExecutor;

    @Inject
    public DataStoreManager(
            RxDataStore<Preferences> dataStore,
            Logger logger,
            @IODispatcher Executor ioExecutor) {
        this.dataStore = dataStore;
        this.logger = logger;
        this.ioExecutor = ioExecutor;
    }

    /**
     * Получает LiveData для значения из DataStore.
     *
     * @param key          Ключ Preferences.
     * @param defaultValue Значение по умолчанию. Для строковых ключей, если вы хотите, чтобы LiveData эмитил null,
     *                     когда ключ отсутствует, передайте сюда специальное не-null значение (например, пустую строку),
     *                     а затем преобразуйте его в null в Transformations.map в вашем репозитории или ViewModel.
     *                     Либо, если тип T это Optional<String>, то можно передавать Optional.empty().
     *                     RxJava map оператор не должен возвращать null.
     * @param <T>          Тип значения.
     * @return LiveData, эмитящий значения.
     */
    public <T> LiveData<T> getPreferenceLiveData(@NonNull Preferences.Key<T> key, @NonNull T defaultValue) {
        Flowable<T> flowable = dataStore.data()
                .map(prefs -> {
                    T value = prefs.get(key);
                    // map не должен возвращать null. Если value null, мы обязаны вернуть defaultValue.
                    // defaultValue здесь должен быть не-null.
                    return value != null ? value : defaultValue;
                })
                .onErrorReturn(error -> {
                    logger.error(TAG, "Error reading preference for key: " + key.getName() + " in LiveData stream. Returning defaultValue.", error);
                    return defaultValue;
                })
                .subscribeOn(Schedulers.from(ioExecutor));

        return LiveDataReactiveStreams.fromPublisher(flowable);
    }


    public <T> ListenableFuture<T> getValueFuture(@NonNull Preferences.Key<T> key, @Nullable T defaultValue) {
        SettableFuture<T> settableFuture = SettableFuture.create();
        Disposable disposable = dataStore.data().firstOrError()
                .map(prefs -> {
                    T value = prefs.get(key);
                    if (value != null) {
                        return value;
                    }
                    // Если значение отсутствует и defaultValue предоставлен (не null), возвращаем его.
                    if (defaultValue != null) {
                        return defaultValue;
                    }
                    // Если значение отсутствует и defaultValue тоже null, Future должен завершиться ошибкой.
                    throw new NoSuchElementException("Key " + key.getName() + " not found and no non-null defaultValue provided.");
                })
                .subscribeOn(Schedulers.from(ioExecutor))
                .subscribe(
                        settableFuture::set,
                        throwable -> {
                            if (throwable instanceof NoSuchElementException) {
                                // Это ожидаемая ошибка, если ключ не найден и defaultValue был null.
                                // Завершаем Future с этой ошибкой.
                                logger.warn(TAG, throwable.getMessage() + " (Future fails with NoSuchElementException).");
                                settableFuture.setException(throwable);
                            } else {
                                logger.error(TAG, "Error getting value for key: " + key.getName() + " with Future.", throwable);
                                settableFuture.setException(new IOException("Failed to get value for " + key.getName(), throwable));
                            }
                        }
                );

        settableFuture.addListener(() -> {
            if (settableFuture.isCancelled() && !disposable.isDisposed()) {
                disposable.dispose();
            }
        }, ioExecutor);
        return settableFuture;
    }
    // ... (saveValueFuture, clearValueFuture без изменений) ...
    public <T> ListenableFuture<Void> saveValueFuture(@NonNull Preferences.Key<T> key, @NonNull T value) {
        SettableFuture<Void> settableFuture = SettableFuture.create();
        Disposable disposable = dataStore.updateDataAsync(prefsIn -> {
                    MutablePreferences mutablePreferences = prefsIn.toMutablePreferences();
                    mutablePreferences.set(key, value);
                    return Single.just(mutablePreferences.toPreferences());
                })
                .subscribeOn(Schedulers.from(ioExecutor))
                .subscribe(
                        prefs -> {
                            logger.debug(TAG, "Value updated in DataStore: " + key.getName() + " -> " + value);
                            settableFuture.set(null);
                        },
                        throwable -> {
                            logger.error(TAG, "Error saving value for key: " + key.getName(), throwable);
                            settableFuture.setException(new IOException("Failed to save value for " + key.getName(), throwable));
                        }
                );
        settableFuture.addListener(() -> {
            if (settableFuture.isCancelled() && !disposable.isDisposed()) {
                disposable.dispose();
            }
        }, ioExecutor);
        return settableFuture;
    }

    public <T> ListenableFuture<Void> clearValueFuture(@NonNull Preferences.Key<T> key) {
        SettableFuture<Void> settableFuture = SettableFuture.create();
        Disposable disposable = dataStore.updateDataAsync(prefsIn -> {
                    MutablePreferences mutablePreferences = prefsIn.toMutablePreferences();
                    mutablePreferences.remove(key);
                    return Single.just(mutablePreferences.toPreferences());
                })
                .subscribeOn(Schedulers.from(ioExecutor))
                .subscribe(
                        prefs -> {
                            logger.debug(TAG, "Value cleared in DataStore: " + key.getName());
                            settableFuture.set(null);
                        },
                        throwable -> {
                            logger.error(TAG, "Error clearing value for key: " + key.getName(), throwable);
                            settableFuture.setException(new IOException("Failed to clear value for " + key.getName(), throwable));
                        }
                );
        settableFuture.addListener(() -> {
            if (settableFuture.isCancelled() && !disposable.isDisposed()) {
                disposable.dispose();
            }
        }, ioExecutor);
        return settableFuture;
    }
}

File: app/src/main/java/com/example/projectquestonjava/core/managers/DataStoreManager.kt
// File: A:\Progects\ProjectQuestOnJava\app\src\main\kotlin\com\example\projectquestonjava\core\managers\DataStoreManager.kt
package com.example.projectquestonjava.core.managers

import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.emptyPreferences
import androidx.lifecycle.LiveData
import androidx.lifecycle.asLiveData
import com.example.projectquestonjava.core.di.IODispatcher
import com.example.projectquestonjava.core.utils.Logger
import com.google.common.util.concurrent.ListenableFuture
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.guava.future
import kotlinx.coroutines.withContext
import java.io.IOException
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class DataStoreManager @Inject constructor(
    private val dataStore: DataStore<Preferences>,
    private val logger: Logger,
    @IODispatcher private val ioDispatcher: CoroutineDispatcher
) {
    private val applicationScope = CoroutineScope(SupervisorJob() + ioDispatcher)

    fun <T : Any> getPreferenceFlow(key: Preferences.Key<T>, defaultValue: T): Flow<T> =
        dataStore.data
            .catch { exception ->
                logger.error("DataStoreManager", "Error reading preferences for key: ${key.name}", exception)
                if (exception is IOException) {
                    emit(emptyPreferences())
                } else {
                    throw exception
                }
            }
            .map { preferences ->
                preferences[key] ?: defaultValue
            }

    fun <T : Any> getPreferenceLiveData(key: Preferences.Key<T>, defaultValue: T): LiveData<T> =
        getPreferenceFlow(key, defaultValue).asLiveData(ioDispatcher)

    suspend fun <T : Any> getValue(key: Preferences.Key<T>, defaultValue: T): T =
        withContext(ioDispatcher) {
            getPreferenceFlow(key, defaultValue).first()
        }

    fun <T : Any> getValueFuture(key: Preferences.Key<T>, defaultValue: T): ListenableFuture<T> {
        return applicationScope.future(ioDispatcher) {
            getPreferenceFlow(key, defaultValue).first()
        }
    }

    suspend fun <T> saveValue(key: Preferences.Key<T>, value: T): Result<Unit> =
        withContext(ioDispatcher) {
            runCatching {
                dataStore.edit { preferences ->
                    preferences[key] = value
                }
                logger.debug("DataStoreManager", "Value updated: ${key.name} -> $value")
            }.onFailure { e ->
                logger.error("DataStoreManager", "Failed to save value for key: ${key.name}", e)
            }
        }

    suspend fun <T> clearValue(key: Preferences.Key<T>): Result<Unit> =
        withContext(ioDispatcher) {
            runCatching {
                dataStore.edit { preferences ->
                    preferences.remove(key)
                }
                logger.debug("DataStoreManager", "Value cleared: ${key.name}")
            }.onFailure { e ->
                logger.error("DataStoreManager", "Failed to clear value for key: ${key.name}", e)
            }
        }

    // --- Java-friendly обертки ---
    @Suppress("UNCHECKED_CAST")
    fun saveValueFuture(key: Preferences.Key<Boolean>, value: Boolean): ListenableFuture<Void> {
        return applicationScope.future(ioDispatcher) {
            saveValue(key, value).getOrThrow()
            null // Явно возвращаем null для Void?
        } as ListenableFuture<Void>
    }

    @Suppress("UNCHECKED_CAST")
    fun saveValueFuture(key: Preferences.Key<Int>, value: Int): ListenableFuture<Void> {
        return applicationScope.future(ioDispatcher) {
            saveValue(key, value).getOrThrow()
            null
        } as ListenableFuture<Void>
    }

    @Suppress("UNCHECKED_CAST")
    fun saveValueFuture(key: Preferences.Key<Long>, value: Long): ListenableFuture<Void> {
        return applicationScope.future(ioDispatcher) {
            saveValue(key, value).getOrThrow()
            null
        } as ListenableFuture<Void>
    }

    @Suppress("UNCHECKED_CAST")
    fun saveValueFuture(key: Preferences.Key<Float>, value: Float): ListenableFuture<Void> {
        return applicationScope.future(ioDispatcher) {
            saveValue(key, value).getOrThrow()
            null
        } as ListenableFuture<Void>
    }

    @Suppress("UNCHECKED_CAST")
    fun saveValueFuture(key: Preferences.Key<String>, value: String): ListenableFuture<Void> {
        return applicationScope.future(ioDispatcher) {
            saveValue(key, value).getOrThrow()
            null
        } as ListenableFuture<Void>
    }

    @Suppress("UNCHECKED_CAST")
    fun saveValueFuture(key: Preferences.Key<Set<String>>, value: Set<String>): ListenableFuture<Void> {
        return applicationScope.future(ioDispatcher) {
            saveValue(key, value).getOrThrow()
            null
        } as ListenableFuture<Void>
    }

    @Suppress("UNCHECKED_CAST")
    fun <T> clearValueFuture(key: Preferences.Key<T>): ListenableFuture<Void> {
        return applicationScope.future(ioDispatcher) {
            clearValue(key).getOrThrow()
            null
        } as ListenableFuture<Void>
    }
}